Layer selector vizjson 3 migration

juanignaciosl opened the pull request:
This closes #10551

juanignaciosl commented:
@javitonino CR this, please.

cc @nobuti

javitonino commented on app/controllers/carto/api/vizjson3_presenter.rb:
```diff
@@ -208,6 +208,20 @@ def html_safe(string)
+
+      def overlays_vizjson(visualization)
+        visualization.overlays.
+          select { |o| !o.type == 'layer_selector' }.
```
`reject`?

javitonino commented on app/controllers/carto/api/vizjson3_presenter.rb:
```diff
@@ -208,6 +208,20 @@ def html_safe(string)
+      def map_options(visualization)
+        map = visualization.map
+
+        migration_options = { layer_selector: true } if visualization.overlays.any? { |o| o.type == 'layer_selector' }
```
I am not sure, but I feel this should be true/false instead of true/missing.

javitonino replied:
I see a potential problem here:
1. Migrate a map with layer selector
2. Disable layer selector in builder
3. Refresh, this presenter will activate the layer selector (although the user has disabled it).

My solution proposal is to remove the overlay in the code that bumps the visualization version on migration.

juanignaciosl replied:
Oh, of course, good catch!

juanignaciosl commented:
@javitonino next CR, please.

juanignaciosl commented:
@javitonino next CR, please. `reload` failed to me during last tests because it's not a sequel method, but I didn't noticed it before. I didn't noticed because I tried migrating only the layer selector, and it doesn't pop up an error message, so request fails silently (but the migration worked). I saw the error when I added unsupported overlays.

I think that it's not actually needed. In addition, I've added error traces to that controller for errors to be properly logged, so I'll ping @CartoDB/builder-backend for everybody to be aware about likely new error traces (not new errors).

javitonino commented on app/controllers/api/json/visualizations_controller.rb:
```diff
@@ -105,7 +105,8 @@ def create
         render_jsonp(vis)
-      rescue CartoDB::InvalidMember
+      rescue CartoDB::InvalidMember => e
+        CartoDB::Logger.error(message: "Invalid member creating visualization", visualization_id: vis.id, exception: e)
```
Is vis.id available here? It seems strange to me as this is a creation.

juanignaciosl replied:
Probably not, because an `InvalidMember` is likely to be thrown before visualization saving, but it might not and it does no harm, as it will be `nil`. Don't you agree?

javitonino replied:
I suppose there is no harm, I was just surprised. :+1:

javitonino commented on lib/carto/visualization_migrator.rb:
```diff
@@ -5,13 +5,30 @@ module VisualizationMigrator
+    def layer_selector_migration(vis)
+      if vis.overlays.any? { |o| o.type == 'layer_selector' }
+        options = vis.map.options.merge(layer_selector: true)
+        vis.map.update_attribute(:options, options)
```
Does `update_attribute` work with Sequel?

juanignaciosl replied:
You're right (although value ends in the column 🤔 ). Probably I haven't tested the "blocked" scenario and a next PUT sets the migrated value, because it worked despite the failure. In fact, `map.options` was nil!

javitonino commented on lib/carto/visualization_migrator.rb:
```diff
@@ -6,8 +6,8 @@ def migrate_visualization_to_v3(vis)
 
-      mapcap_creation(vis)
       layer_selector_migration(vis)
+      mapcap_creation(vis)
```
Oh, good catch with the reorder! :+1:

juanignaciosl commented:
@javitonino last? CR, please

javitonino commented on app/models/map.rb:
```diff
@@ -6,6 +6,8 @@
 class Map < Sequel::Model
   self.raise_on_save_failure = false
 
+  plugin :serialization, :json, :options
```
This deserializes keys as strings. But when merging the options dictionary later, you merge using a symbols as a key. I tested, and what happens is that the merged hash has two keys `"layer_selector"` and `:layer_selector`. In my test, when serializing back (in `save`) it keeps the symbol key, but I do not know if that's trustable.

juanignaciosl replied:
:-\ Could you give me STR, please? I tested it and I didn't get any duplicated key :(

juanignaciosl replied:
Anyway, this is dirty. I'll try to use only strings (and do a "_unit_" test for the migrator).

javitonino replied:
Be careful because `Carto::Map` serialization uses symbols :cry:

juanignaciosl replied:
(╯°□°）╯︵ ┻━┻

javitonino commented on lib/carto/visualization_migrator.rb:
```diff
@@ -5,13 +5,36 @@ module VisualizationMigrator
+          map.update_attribute(:options, options)
+        else
+          map.options = options
+          map.save
```
This second branch should work for both ORMs

juanignaciosl commented:
@javitonino CR, please U_U. I haven't changed anything related to key/symbols, because I think that it works. You can test it in ded06, let me know if you manage to break it, please. cc @nobuti

juanignaciosl commented:
Thanks!
